### ISSUE TITLE
proposal: net: add support for Happy Eyeballs version 2 (RFC 8305)

### ISSUE THREAD

**Issue by mikioh** [golang/go#23841]
Since RFC 6555 published, research papers, for example, Measuring the Effects of Happy Eyeballs, depict improvement space for the RFC. In addition, network or service operators speak up that the race is not only for TCP connection setup but for DNS query handling. [RFC 8305](https://tools.ietf.org/html/rfc8305) Happy Eyeballs version 2 addresses such issues.

To summarize, version 2 clarifies the existing algorithm, modifies the configurable variable values for performance improvement and adds consideration for various circumstances such as IPv6-IPv4 transition and VPN. Luckily, there's no significant user-visible behavioral change.

I'd like to propose to replace the existing RFC 6555 implementation with RFC 8305 implementation after landing dns/dnsmessage package (#16218).

**Comment by bradfitz** [comment]
Summarize what this means for Go, so we don't need to diff RFC 6555 vs 8305.

Just add Happy Eyeballs-ish behavior to DNS lookups?

**Comment by mikioh** [comment]
> Just add Happy Eyeballs-ish behavior to DNS lookups?

Yup, mostly. FWIW, https://tools.ietf.org/html/rfc8305#appendix-A

**Comment by bradfitz** [comment]
Seems fine. As long as the code can be done without too much complexity.

**Comment by rsc** [comment]
Accepted but only if the result is simple enough.

**Comment by fortuna** [comment]
Happy Eyeballs v2 can provide a meaningful improvement in latency. I have a mostly compliant Go implementation here:
https://pkg.go.dev/github.com/Jigsaw-Code/outline-sdk/transport#HappyEyeballsStreamDialer.

The only part missing is sorting IPs as per https://datatracker.ietf.org/doc/html/rfc8305#section-4, but it's been implemented for Happy Eyeballs v1 in the standard library and we can borrow from that.

**Comment by seankhliao** [comment]
feel free to send a CL

**Comment by fortuna** [comment]
I may at some point if I free up some time. I just wanted to register the implementation I have in case anyone wants to explore this sooner.
